File: P1/Ejercicio4/core/filter_chain.rb
class FilterChain

  # @filters almacena todos los filtros a ejecutar
  # @target es el servicio final
  def initialize
    @filters = []
    @target = nil
  end

  # añade un nuevo filtro a la cadena
  # los filtros se ejecutarán en el orden en que se añaden
  def add_filter(filter)
    @filters << filter
  end

  # establece el servicio final
  def set_target(target)
    @target = target
  end

  # ejecuta la cadena de filtros
  # cada filtro valida la petición; si alguno falla, lanza excepción
  # y se detiene la ejecución
  # si todos los filtros pasan, se ejecuta el target
  def execute(request)
    @filters.each do |filter|
      filter.execute(request)
    end

    # se ejecuta el servicio final solo si todos los filtros han pasado
    @target.execute(request)
  end

end

File: P2/Ejercicio1_MantenimientoFlutter/lib/core/FilterChain.dart
import '../interfaces/Filter.dart';
import 'AuthenticationService.dart';

class FilterChain {
  final List<Filter> _filters = [];
  AuthenticationService? _target;

  void addFilter(Filter filter) {
    _filters.add(filter);
  }

  void setTarget(AuthenticationService target) {
    _target = target;
  }

  // devuelve un String con el error si algo falla, 
  // o null si todo salió bien y se ejecutó el target
  String? execute(Map<String, String> request) {
    for (var filter in _filters) {
      String? errorMessage = filter.execute(request);
      
      if (errorMessage != null) {
        // si un filtro falla, devolvemos el mensaje inmediatamente
        return errorMessage; 
      }
    }

    // si todos los filtros pasaron (devolvieron null), ejecutamos el target
    _target?.execute(request);
    return null; 
  }
}